tfhe/benches/integer/bench.rs

```diff
@@ -2,138 +2,169 @@
+        if self.indices.1 >= self.bit_sizes.len() {
+            self.indices.1 = 0;
+            self.indices.0 += 1;
+        }
 
-        let clear_0 = rng.gen::<u64>() % modulus;
+        if self.indices.0 >= self.params.len() {
+            return None;
+        }
 
-        let mut ct = cks.encrypt_radix(clear_0, param.num_block);
+        let param = self.params[self.indices.0];
+        let bit_size = self.bit_sizes[self.indices.1];
+        let num_block =
+            (bit_size as f64 / (param.message_modulus.0 as f64).log(2.0)).ceil() as usize;
 
-        let bench_id = param_name;
-        bench_group.bench_function(bench_id, |b| {
-            b.iter(|| {
-                sks.smart_neg(&mut ct);
-            })
-        });
+        self.indices.1 += 1;
+        Some((param, num_block, bit_size))
     }
-
-    bench_group.finish()
 }
 
-fn full_propagate(c: &mut Criterion) {
-    let mut bench_group = c.benchmark_group("full_propagate");
-
-    for (param_name, param) in SERVER_KEY_BENCH_PARAMS {
-        let (cks, sks) = KEY_CACHE.get_from_params(param.block_parameters);
-        let mut rng = rand::thread_rng();
-
-        let modulus = (param.block_parameters.message_modulus.0 * param.num_block) as u64;
+/// Base function to bench a server key function that is a binary operation
+fn bench_server_key_binary_function<F>(c: &mut Criterion, bench_name: &str, binary_op: F)
+where
+    F: Fn(&ServerKey, &mut RadixCiphertext, &mut RadixCiphertext),
+{
+    let mut bench_group = c.benchmark_group(bench_name);
+    bench_group
+        .sample_size(15)
+        .measurement_time(std::time::Duration::from_secs(60));
+    let mut rng = rand::thread_rng();
 
-        let clear_0 = rng.gen::<u64>() % modulus;
+    for (param, num_block, bit_size) in ParamsAndNumBlocksIter::default() {
+        let param_name = param.name();
 
-        let mut ct = cks.encrypt_radix(clear_0, param.num_block);
+        let bench_id = format!("{param_name}/{bit_size}_bits");
+        bench_group.bench_function(&bench_id, |b| {
+            let (cks, sks) = KEY_CACHE.get_from_params(param);
 
-        let bench_id = param_name;
-        bench_group.bench_function(bench_id, |b| {
-            b.iter(|| {
-                sks.full_propagate(&mut ct);
-            })
+            let encrypt_two_values = || {
+                let clearlow = rng.gen::<u128>();
+                let clearhigh = rng.gen::<u128>();
+                let clear_0 = tfhe::integer::U256::from((clearlow, clearhigh));
+                let mut ct_0 = cks.encrypt_radix(clear_0, num_block);
+
+                let clearlow = rng.gen::<u128>();
+                let clearhigh = rng.gen::<u128>();
+                let clear_1 = tfhe::integer::U256::from((clearlow, clearhigh));
+                let mut ct_1 = cks.encrypt_radix(clear_1, num_block);
+
+                // Raise the degree, so as to ensure worst case path in operations
+                let mut carry_mod = param.carry_modulus.0;
```

Comment:
To me it makes since to have benches for the worst case, a bit like there is the big O notation for worst case algorithm complexity

But the question is more general how do we want to bench smart operations

Reply:
Yep I don't have a strong opinion on this, worth a discussion as a team/with benchmarking people :)